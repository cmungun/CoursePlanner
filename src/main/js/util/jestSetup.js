import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });
// Make Enzyme functions available in all test files without importing
global.shallow = shallow;
global.render = render;
global.mount = mount;

global.cloneObject = (obj) => {
    return JSON.parse(JSON.stringify(obj));
};

// prevents ajax requests (and any other jquery functions) from actually firing
jest.mock('jquery');