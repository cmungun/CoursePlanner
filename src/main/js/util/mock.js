/*
 *  Add mock data to be used in unit tests to this file.
 *  Never remove or change the contents of the variables. Only add new content.
 *
 */

export const mockPositionObjects = [
    {
        season: "fall",
        yearIndex: 0,
        courseIndex: 0
    },
    {
        season: "fall",
        yearIndex: 0,
        courseIndex: 0,
        orListIndex: 0
    }
];

export const mockSearchBoxInput = [
    "com",
    "comp 24",
    "comp 249",
    "COMP 249"
];