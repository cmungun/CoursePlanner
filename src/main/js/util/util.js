/*
 *  Collection of utilities such as constants or functions that may be used by more than one component
 *
 *  All constants should be kept in this file.
 *
 */
import React from "react";
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import CircularProgress from 'material-ui/CircularProgress';

// App theme

export const MUI_THEME = getMuiTheme({
    palette: {
        primary1Color: "#6c1540",
        primary2Color: "#3d001a",
        accent1Color: "#f5bb2b",
        pickerHeaderColor: "#6c1540",
    }
});

// Regular old constants
export const SEASON_NAMES_PRETTY = ["Fall", "Winter", "Summer"];
export const SEASON_NAMES = SEASON_NAMES_PRETTY.map((season) => season.toLowerCase());
export const MAX_UNDO_HISTORY_LENGTH = 100;
export const AUTO_SCROLL_PAGE_PORTION = 0.1; // auto scroll on the top and bottom 10% of screen
export const AUTO_SCROLL_STEP = 10;
export const AUTO_SCROLL_DELAY = 20;
export const FEEDBACK_CHAR_LIMIT = 1000;
export const FEEDBACK_SNACKBAR_AUTOHIDE_DURATION = 4000;
export const FEEDBACK_ROWS_MAX = 10;

export const KEY_CODES = {
    SHIFT: 16,
    CTRL: 17,
    Z: 90
};

/*
 * All hardcoded pieces of text which are directly displayed to the user.
 * Add '_T' suffix to property name if it uses sprintf template syntax
 */
export const UI_STRINGS = {

    SITE_NAME: "ConU Course Planner",
    BETA_LABEL: "Beta",

    PROGRAM_SELECTION_LOADING: "Loading list of recommended sequences",
    PROGRAM_SELECTION_TITLE: "Welcome to ConU Course Planner!",
    PROGRAM_SELECTION_PROGRAM_TITLE: "Select your program",
    PROGRAM_SELECTION_OPTION_TITLE: "Select your option",
    PROGRAM_SELECTION_ENTRY_TYPE_TITLE: "Select your entry type",
    PROGRAM_SELECTION_FINAL_MESSAGE: "Your selected program is:",

    PROGRAM_SELECTION_CONFIRM_LABEL: "Confirm",
    PROGRAM_SELECTION_BACK_LABEL: "Back",
    PROGRAM_SELECTION_PREVIOUS_LABEL: "Previous",
    PROGRAM_SELECTION_FINISH_LABEL: "Finish",
    PROGRAM_SELECTION_NEXT_LABEL: "Next",

    WORK_TERM: "Work Term",
    IS_WORK_TERM: "is work term?",
    NO_COURSES: "No Courses",

    DEFAULT_SEARCH_LABEL: "Course search by course code",
    COURSE_SEARCH_HINT: "e.g. comp 248",
    COURSE_SEARCH_FOUND_NONE: "Search returned no results",

    COURSE_INFO_LOADING: "Getting course info",

    COURSE_INFO_HEADING_DESCRIPTION: "Description",
    COURSE_INFO_HEADING_PREREQUISITES: "Pre-requisites",
    COURSE_INFO_HEADING_COREQUISITES: "Co-requisites",

    COURSE_INFO_HINT: "Click on or search for a course to display its info",
    VALIDATION_RESULTS_HINT: "Make changes to your schedule to trigger a validation",

    VALIDATION_LOADING: "Validating course sequence",
    
    VALIDATION_FAILURE_MSG: "Sequence contains issues/warnings",
    VALIDATION_SUCCESS_MSG: "Sequence is valid",

    VALIDATION_MISSING_REQUISITE_T: "%1$s is missing %2$s: %3$s",
    VALIDATION_MISSING_CREDITS_T: "Sequence contains only %1$s of %2$s required credits",
    VALIDATION_REPEATED_COURSE_T: "%1$s is repeated %2$s times in the sequence",
    VALIDATION_NO_OPTION_SELECT_T: "No option selected in %1$s of year %2$s, entry number %3$s",

    EXPORT_TEXT: "Export",
    EXPORTING_SEQUENCE: "Exporting sequence",

    SELECT_NEW_PROGRAM: "Select a new program",

    ELECTIVE_COURSE_TOOLTIP: "Replace me with a real course!",
    ORLIST_CHOICE_TOOLTIP: "Choose course from list of options",
    LIST_LOADING: "Loading...",
    LIST_NONE_SELECTED: "None Selected",

    REPO_LINK_TEXT: "View on Github",

    FEEDBACK_TEXT: "Feedback",
    FEEDBACK_CHAR_LIMIT_ERROR_MSG: "Character limit reached",
    FEEDBACK_PLACEHOLDER: "Insert Feedback here",
    FEEDBACK_BOX_TITLE: "Comments/Concerns? Let us know and we'll see what we can do",
    FEEDBACK_SEND_ERROR_MSG: "An error occurred, message not sent",
    FEEDBACK_SEND_SUCCESS_MSG: "Message sent, thank you for your feedback!"
};

export const URLS = {
    REPO: "https://github.com/stumash/CoursePlanner"
};

export const PROGRAM_NAMES = {
    SOEN: "Software Engineering",
    COMP: "Computer Science",
    BLDG: "Building Engineering",
    CIVI: "Civil Engineering",
    INDU: "Industrial Engineering",
    MECH: "Mechanical Engineering",
    COEN: "Computer Engineering",
    ELEC: "Electrical Engineering",
    AERO: "Aerospace Engineering"
};

export const PROGRAM_OPTIONS = {
    General: "General Program",
    Games: "Computer Games",
    Realtime: "Real-time, Embedded and Avionics Software ",
    Web: "Web Services and Applications",
    Apps: "Computer Applications",
    CompSys: "Computer Systems",
    InfoSys: "Information Systems",
    Stats: "Mathematics and Statistics",
    SoftSys: "Software Systems",
    CompArts: "Computation Arts",
    NoOption: "None",
    Tele: "Telecommunications",
    Electronics: "Electronics/VLSI",
    Avionics: "Avionics and Control Systems",
    Power: "Power and Renewable Energy",
    Propulsion: "Aerodynamics and Propulsion",
    Structures: "Aerospace Structures and Materials",
    AeroSys: "Avionics and Aerospace Systems"
};

export const PROGRAM_ENTRY_TYPES = {
    Sept: "September",
    Jan: "January",
    Coop: "Coop September"
};

export const EXPORT_TYPES = {
    EXPORT_TYPE_TABLE_PDF: "PDF (table)",
    EXPORT_TYPE_TABLE_HTML: "HTML (table)",
    EXPORT_TYPE_LIST_PDF: "PDF (list)",
    EXPORT_TYPE_LIST_MD: "TEXT (list)"
};

export const COURSE_EXEMPTIONS = [
    "MATH 203",
    "MATH 204",
    "MATH 205",
    "CHEM 205",
    "PHYS 204",
    "PHYS 205"
];

export const INLINE_STYLES = {
    appBar: {
        zIndex: "0"
    },
    appBarVertIcon: {
        color: "white"
    },
    exportLoadingDialogTitle: {
        textAlign: "center"
    },
    exportLoadingDialogContent: {
        width: "300px"
    },
    exportLoadingDialogIcon: {
        width: "100%",
        textAlign: "center"
    },
    autoCompleteContainer: {
        marginLeft: "12px"
    },
    autoCompleteList: {
        maxHeight: "250px",
        overflow: "auto"
    },
    searchLoadingIcon: {
        paddingLeft: "12px"
    },
    courseInfoLoadingIcon: {
        marginLeft: "-8px",
        marginTop: "-10px"
    },
    courseInfoListItem: {
        padding: "8px 0",
        height: "32px",
        marginLeft: "16px",
        fontSize: "14px"
    },
    courseInfoCardText: {
        paddingTop: "0"
    },
    validationLoadingIcon: {
        marginLeft: "-8px",
        marginTop: "-10px"
    },
    allSequencesLoadingIcon: {
        width: "100%",
        textAlign: "center"
    },
    programSelectContent: {
        width: "40%"
    },
    programSelectNextButton: {
        marginRight: "12px"
    },
    betterYellow: {
        color: "#f5bb2b"
    }
};

export const SEMESTER_ITEM_CLASS_MAP = {
    isHidden: "isHidden",
    isDraggable: "grabbable",
    isHighlighted: "highlighted",
    isSelected: "selected",
    isDragPreview: "dragPreview"
};

export const LOADING_ICON_TYPES = {
    big: <CircularProgress size={80} thickness={7}/>,
    small: <CircularProgress size={25} thickness={2.5}/>,
    export: <CircularProgress size={80} thickness={7} style={INLINE_STYLES.exportLoadingDialogIcon}/>
};

export function generatePrettyProgramName(program, option, entryType, minTotalCredits){
    return (
        PROGRAM_NAMES[program] + ", " +
        ((option !== "NoOption") ? PROGRAM_OPTIONS[option] + " Option, " : "") +
        PROGRAM_ENTRY_TYPES[entryType] + " Entry" +
        (minTotalCredits ?  " (" + minTotalCredits + " credits)" : "")
    );
};

/*
 *  Convenience function that allows you to save the file contained at location uri to disk of client machine
 *  currently used for downloading exported PDF file
 */
export function saveAs(uri, filename) {
    const link = document.createElement('a');
    if (typeof link.download === 'string') {
        document.body.appendChild(link); // Firefox requires the link to be in the body
        link.download = filename;
        link.href = uri;
        link.click();
        document.body.removeChild(link); // remove the link when done
    } else {
        location.replace(uri);
    }
}

/*
 *  For every course and orList in the yearList, we must give it a unique ID.
 *  This is helpful for react to properly respond to changes and
 *  it is especially necessary for the react-dnd module to work properly
 *
 */
export function generateUniqueKeys(yearList){
    yearList.forEach((year, yearIndex) => {
        SEASON_NAMES.forEach((season) => {
            year[season].courseList.forEach((courseObj, courseIndex) => {
                if(courseObj.length > 0){
                    courseObj.forEach((courseOrObj, orListIndex) => {
                        courseOrObj.id = generateUniqueKey(courseOrObj, season, yearIndex, courseIndex, orListIndex);
                    });
                } else {
                    courseObj.id = generateUniqueKey(courseObj, season, yearIndex, courseIndex, "");
                }
            });
        });
    });
    return yearList;
}

/*
 *  Form unique ID by combing course code/electiveType with its current position in the yearList
 *  If the course changes its position within the yearList, we do NOT want this id value to change, so we only call this once.
 */
export function generateUniqueKey(courseObj, season, yearIndex, courseIndex, orListIndex, timestamp){
    let id = (courseObj.isElective === "true") ? courseObj.electiveType : courseObj.code;
    id += positionToString({
        season: season,
        yearIndex: yearIndex,
        courseIndex: courseIndex,
        orListIndex: orListIndex
    });
    (timestamp) && (id += (" " + timestamp));
    return id;
}

/*
 * Generate a string from a position object
 *     position: position of the course in the sequence, possibly within an orList
 */
export function positionToString(position){
    return [position.season, position.yearIndex, position.courseIndex].join(" ");
}

/*
 * Generate a position from a position string
 *     positionString: position of the course in the sequence, represented as a string
 */
export function parsePositionString(positionString){
    let subStrings = positionString.split(" ");
    return {
        season: subStrings[0],
        yearIndex: parseInt(subStrings[1]),
        courseIndex: parseInt(subStrings[2])
    };
}

// Item types used for DND
export const ITEM_TYPES = {
    COURSE: "Course",
    OR_LIST: "OrList"
};

/*
 *  Special object used by react-dnd to register a drag source
 */
export function createDragSourceSpec(itemType) {
    let beginDrag = undefined;
    if(itemType === ITEM_TYPES.COURSE) {
        beginDrag = (props, monitor, component) => {
            props.onChangeDragState && props.onChangeDragState(true, props.position, props.courseObj);
            return {
                courseObj: props.courseObj,
                position: props.position
            };
        }
    }
    if(itemType === ITEM_TYPES.OR_LIST) {
        beginDrag = (props, monitor, component) => {
            props.onChangeDragState && props.onChangeDragState(true, props.position);
            return {
                courseList: props.courseList,
                position: props.position
            };
        }
    }
    return {
        beginDrag: beginDrag,
        endDrag(props, monitor, component){
            props.onChangeDragState && props.onChangeDragState(false, props.position);
        },
        canDrag(props, monitor){
            return props.isDraggable;
        }
    };
}

/*
 *  Collect function used by react-dnd to inject properties into a drag source
 */
export function collectSource(connect, monitor) {
    return {
        connectDragSource: connect.dragSource(),
        isBeingDragged: monitor.isDragging()
    };
}