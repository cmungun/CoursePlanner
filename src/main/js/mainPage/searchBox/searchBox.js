import React from "react";
import AutoComplete from 'material-ui/AutoComplete';

import {UI_STRINGS, INLINE_STYLES, LOADING_ICON_TYPES} from "../../util/util";

import $ from "jquery";

/*
 *  Text input which is used to filter/search through course codes in DB
 *
 *  Expects props:
 *
 *  onConfirmSearch - see MainPage.loadCourseInfo
 *
 */
export class SearchBox extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            isFiltering: false,
            filterResults: [],
            floatingLabelText: UI_STRINGS.DEFAULT_SEARCH_LABEL
        };

        // functions that are passed as callbacks need to be bound to current class - see https://facebook.github.io/react/docs/handling-events.html
        this.filterCourseCodes = this.filterCourseCodes.bind(this);
        this.onFilterSuccess = this.onFilterSuccess.bind(this);
        this.handleNewRequest = this.handleNewRequest.bind(this);
    }

    handleNewRequest(chosenRequest, index) {
        if(index === -1){
            this.props.onConfirmSearch(chosenRequest.toUpperCase());
        } else if(chosenRequest.value.length > 0) {
            this.props.onConfirmSearch(chosenRequest.value.toUpperCase());
        }
    }

    render() {
        return (
            <div className="courseSearch">
                <AutoComplete id="courseSearch"
                              filter={AutoComplete.noFilter}
                              hintText={UI_STRINGS.COURSE_SEARCH_HINT}
                              dataSource={this.state.filterResults}
                              floatingLabelText={this.state.floatingLabelText}
                              style={INLINE_STYLES.autoCompleteContainer}
                              listStyle={INLINE_STYLES.autoCompleteList}
                              onUpdateInput={this.filterCourseCodes}
                              onNewRequest={this.handleNewRequest}/>
                {this.state.isFiltering && LOADING_ICON_TYPES.small}
            </div>
        );
    }

    /*
    *  Backend API calls:
    */

    filterCourseCodes(query){
        this.setState({ isFiltering: true });
        $.ajax({
            type: "POST",
            url: "api/filtercoursecodes",
            data: JSON.stringify({ filter : query }),
            success: this.onFilterSuccess
        });
    }

    onFilterSuccess(responseText) {
        let response = JSON.parse(responseText);
        let floatingLabelText = (response.length === 0) ? UI_STRINGS.COURSE_SEARCH_FOUND_NONE : UI_STRINGS.DEFAULT_SEARCH_LABEL;
        let results = response.map(courseCode => ({ text: courseCode, value: courseCode }));
        this.setState({isFiltering: false, filterResults: results, floatingLabelText: floatingLabelText});
    }
}