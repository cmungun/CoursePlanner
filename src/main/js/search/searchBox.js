import React from "react";
import AutoComplete from 'material-ui/AutoComplete';

import {UI_STRINGS, INLINE_STYLES, LOADING_ICON_TYPES} from "../util/util";

/*
 *  Text input which is used to filter/search through course codes in DB
 *
 *  Expects props:
 *
 *  onConfirmSearch - see MainPage.loadCourseInfo
 *
 */
export class SearchBox extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            "isFiltering": false,
            "dataSource": [],
            "floatingLabelText": UI_STRINGS.DEFAULT_SEARCH_LABEL
        };

        // functions that are passed as callbacks need to be bound to current class - see https://facebook.github.io/react/docs/handling-events.html
        this.filterCourseCodes = this.filterCourseCodes.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleNewRequest = this.handleNewRequest.bind(this);
    }

    handleNewRequest(chosenRequest, index) {
        if(index === -1){
            this.props.onConfirmSearch(chosenRequest.toUpperCase());
        } else if(chosenRequest.value.length > 0) {
            this.props.onConfirmSearch(chosenRequest.value.toUpperCase());
        }
    }

    handleInputChange(searchText){
        this.filterCourseCodes(searchText);
    }

    render() {
        return (
            <div className="courseSearch">
                <AutoComplete filter={AutoComplete.noFilter}
                              hintText={UI_STRINGS.COURSE_SEARCH_HINT}
                              dataSource={this.state.dataSource}
                              floatingLabelText={this.state.floatingLabelText}
                              style={INLINE_STYLES.autoCompleteContainer}
                              listStyle={INLINE_STYLES.autoCompleteList}
                              onUpdateInput={this.handleInputChange}
                              onNewRequest={this.handleNewRequest}/>
                {this.state.isFiltering && LOADING_ICON_TYPES.small}
            </div>
        );
    }

    /*
    *  Backend API calls:
    */

    filterCourseCodes(query){
        this.setState({ "isFiltering": true});
        $.ajax({
            type: "POST",
            url: "api/filtercoursecodes",
            data: JSON.stringify({"filter" : query}),
            success: (res) => {
                let response = JSON.parse(res);
                let floatingLabelText = (query.length > 0 && response.length === 0) ? UI_STRINGS.COURSE_SEARCH_FOUND_NONE : UI_STRINGS.DEFAULT_SEARCH_LABEL;
                let newDataSource = [];
                response.forEach((courseCode) => {
                    newDataSource.push({text: courseCode, value: courseCode});
                });
                this.setState({"isFiltering": false, dataSource: newDataSource, floatingLabelText: floatingLabelText});
            }
        });
    }

}