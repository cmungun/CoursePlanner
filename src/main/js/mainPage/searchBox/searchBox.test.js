import React from "react";
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import $ from "jquery";
import { SearchBox } from "./searchBox";
import { mockSearchBoxInput, mockSearchResults } from "../../util/mock";
import { MUI_THEME, UI_STRINGS } from "../../util/util";

let searchBoxClass, mockOnConfirmSearch;

beforeEach(() => {
    $.ajax.mockClear();
    mockOnConfirmSearch = jest.fn();
    searchBoxClass = new SearchBox({
        onConfirmSearch: mockOnConfirmSearch
    });
});

describe("handleNewRequest", () => {
    test("if index is -1, should call onConfirmSearch with the chosenRequest in upper case", () => {
        let chosenRequest = cloneObject(mockSearchBoxInput[0]);
        let index = -1;
        searchBoxClass.handleNewRequest(chosenRequest, index);
        expect(mockOnConfirmSearch.mock.calls[0][0]).toBe(chosenRequest.toUpperCase());
    });

    describe("if index is not -1", () => {
        let index = 0;

        test("if the chosenRequest has a non-empty value, should call onConfirmSearch with the value in upper case", () => {
            let chosenRequest = {value: cloneObject(mockSearchBoxInput[0])};
            searchBoxClass.handleNewRequest(chosenRequest, index);
            expect(mockOnConfirmSearch.mock.calls.length).toBe(1);
            expect(mockOnConfirmSearch.mock.calls[0][0]).toBe(chosenRequest.value.toUpperCase());
        });

        test("if the chosenRequest has an empty value, should not call onConfirmSearch", () => {
            let chosenRequest = {value: ""};
            searchBoxClass.handleNewRequest(chosenRequest, index);
            expect(mockOnConfirmSearch.mock.calls.length).toBe(0);
        });
    });
});

describe("filterCourseCodes", () => {
    beforeEach(() => {
        searchBoxClass.setState = jest.fn();
    });

    test("should set the isFiltering state variable to true", () => {
        let searchQuery = cloneObject(mockSearchBoxInput[0]);
        searchBoxClass.filterCourseCodes(searchQuery);
        expect(searchBoxClass.setState.mock.calls.length).toBe(1);
        expect(searchBoxClass.setState.mock.calls[0][0]).toEqual({ isFiltering: true });
    });

    test("should perform a POST request at api/filtercoursecodes", () => {
        let searchQuery = cloneObject(mockSearchBoxInput[0]);
        searchBoxClass.filterCourseCodes(searchQuery);

        let mockCalls = $.ajax.mock.calls;
        expect(mockCalls.length).toBe(1);

        let ajaxArgs = mockCalls[0][0];
        expect(ajaxArgs.type).toBe("POST");
        expect(ajaxArgs.url).toBe("api/filtercoursecodes");
        expect(ajaxArgs.data).toBe(JSON.stringify({ filter: searchQuery }));
        expect(ajaxArgs.success).toBe(searchBoxClass.onFilterSuccess);
    });
});

describe("onFilterSuccess", () => {
    let mockSearchResultsClone, mockCalls, args;

    let performTest = (results) => {
        searchBoxClass.onFilterSuccess(JSON.stringify(results));
        mockCalls = searchBoxClass.setState.mock.calls;
        expect(mockCalls.length).toBe(1);
        args = mockCalls[0][0];
    };

    beforeEach(() => {
        searchBoxClass.setState = jest.fn();
        mockSearchResultsClone = cloneObject(mockSearchResults);
    });

    test("should update the isFiltering, filterResults, and floatingLabelText state variables", () => {
        performTest(mockSearchResultsClone);
        expect(Object.keys(args)).toEqual(["isFiltering", "filterResults", "floatingLabelText"]);
    });

    test("should set the isFiltering state variable to false", () => {
        performTest(mockSearchResultsClone);
        expect(args.isFiltering).toBe(false);
    });

    describe("if the response text is an empty array", () => {
        beforeEach(() => {
            performTest([]);
        });

        test("should set the floating label text to UI_STRINGS.COURSE_SEARCH_FOUND_NONE", () => {
            expect(args.floatingLabelText).toBe(UI_STRINGS.COURSE_SEARCH_FOUND_NONE);
        });

        test("should set filterResults to an empty array", () => {
            expect(args.filterResults).toEqual([]);
        });
    });

    describe("if the response text is an array with length greater than 0", () => {
        beforeEach(() => {
            expect(mockSearchResultsClone.length).toBeGreaterThan(0);
            performTest(mockSearchResultsClone);
        });

        test("should set the floating label text to UI_STRINGS.DEFAULT_SEARCH_LABEL", () => {
            expect(args.floatingLabelText).toBe(UI_STRINGS.DEFAULT_SEARCH_LABEL);
        });

        test("should add an object with a text and value property to the filterResults array for each search result", () => {
            expect(args.filterResults.length).toBe(mockSearchResultsClone.length);
            expect(Object.keys(args.filterResults[0])).toEqual(["text", "value"]);
        });
    });
});

describe("DOM", () => {
    let searchBoxWrapper, searchBoxInstance;

    beforeEach(() => {
        searchBoxWrapper = shallow(
            <MuiThemeProvider muiTheme={MUI_THEME}>
                <SearchBox onConfirmSearch={jest.fn()}/>
            </MuiThemeProvider>
        ).dive();
        searchBoxInstance = searchBoxWrapper.instance();
    });

    test("should display a loading icon while filtering and display the results after filtering", () => {
        // make snapshot of inital DOM
        expect(searchBoxWrapper).toMatchSnapshot();

        // make snapshot of DOM with loading animation
        let searchQuery = cloneObject(mockSearchBoxInput[0]);
        searchBoxInstance.filterCourseCodes(searchQuery);
        searchBoxWrapper.update();
        expect(searchBoxWrapper).toMatchSnapshot();

        // make snapshot of DOM without loading animation
        let searchResults = cloneObject(mockSearchResults);
        searchBoxInstance.onFilterSuccess(JSON.stringify(searchResults));
        searchBoxWrapper.update();
        expect(searchBoxWrapper).toMatchSnapshot();
    });
});