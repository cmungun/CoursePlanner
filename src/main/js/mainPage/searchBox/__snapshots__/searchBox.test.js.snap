// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM should display a loading icon while filtering and display the results after filtering 1`] = `
<div
  className="courseSearch"
>
  <AutoComplete
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    animated={true}
    dataSource={Array []}
    dataSourceConfig={
      Object {
        "text": "text",
        "value": "value",
      }
    }
    disableFocusRipple={true}
    filter={[Function]}
    floatingLabelText="Course search by course code"
    fullWidth={false}
    hintText="e.g. comp 248"
    id="courseSearch"
    listStyle={
      Object {
        "maxHeight": "250px",
        "overflow": "auto",
      }
    }
    menuCloseDelay={300}
    onNewRequest={[Function]}
    onUpdateInput={[Function]}
    open={false}
    openOnFocus={false}
    style={
      Object {
        "marginLeft": "12px",
      }
    }
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  />
</div>
`;

exports[`DOM should display a loading icon while filtering and display the results after filtering 2`] = `
<div
  className="courseSearch"
>
  <AutoComplete
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    animated={true}
    dataSource={Array []}
    dataSourceConfig={
      Object {
        "text": "text",
        "value": "value",
      }
    }
    disableFocusRipple={true}
    filter={[Function]}
    floatingLabelText="Course search by course code"
    fullWidth={false}
    hintText="e.g. comp 248"
    id="courseSearch"
    listStyle={
      Object {
        "maxHeight": "250px",
        "overflow": "auto",
      }
    }
    menuCloseDelay={300}
    onNewRequest={[Function]}
    onUpdateInput={[Function]}
    open={false}
    openOnFocus={false}
    style={
      Object {
        "marginLeft": "12px",
      }
    }
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  />
  <CircularProgress
    max={100}
    min={0}
    mode="indeterminate"
    size={25}
    thickness={2.5}
    value={0}
  />
</div>
`;

exports[`DOM should display a loading icon while filtering and display the results after filtering 3`] = `
<div
  className="courseSearch"
>
  <AutoComplete
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    animated={true}
    dataSource={
      Array [
        Object {
          "text": "COMM 210",
          "value": "COMM 210",
        },
        Object {
          "text": "COMM 222",
          "value": "COMM 222",
        },
        Object {
          "text": "COMM 223",
          "value": "COMM 223",
        },
        Object {
          "text": "COMM 225",
          "value": "COMM 225",
        },
        Object {
          "text": "COMM 308",
          "value": "COMM 308",
        },
        Object {
          "text": "COMP 108",
          "value": "COMP 108",
        },
        Object {
          "text": "COMP 201",
          "value": "COMP 201",
        },
        Object {
          "text": "COMP 208",
          "value": "COMP 208",
        },
        Object {
          "text": "COMP 218",
          "value": "COMP 218",
        },
        Object {
          "text": "COMP 228",
          "value": "COMP 228",
        },
        Object {
          "text": "COMP 232",
          "value": "COMP 232",
        },
        Object {
          "text": "COMP 233",
          "value": "COMP 233",
        },
        Object {
          "text": "COMP 248",
          "value": "COMP 248",
        },
        Object {
          "text": "COMP 249",
          "value": "COMP 249",
        },
        Object {
          "text": "COMP 326",
          "value": "COMP 326",
        },
        Object {
          "text": "COMP 335",
          "value": "COMP 335",
        },
        Object {
          "text": "COMP 339",
          "value": "COMP 339",
        },
        Object {
          "text": "COMP 345",
          "value": "COMP 345",
        },
        Object {
          "text": "COMP 346",
          "value": "COMP 346",
        },
        Object {
          "text": "COMP 348",
          "value": "COMP 348",
        },
        Object {
          "text": "COMP 352",
          "value": "COMP 352",
        },
        Object {
          "text": "COMP 353",
          "value": "COMP 353",
        },
        Object {
          "text": "COMP 354",
          "value": "COMP 354",
        },
        Object {
          "text": "COMP 361",
          "value": "COMP 361",
        },
        Object {
          "text": "COMP 367",
          "value": "COMP 367",
        },
      ]
    }
    dataSourceConfig={
      Object {
        "text": "text",
        "value": "value",
      }
    }
    disableFocusRipple={true}
    filter={[Function]}
    floatingLabelText="Course search by course code"
    fullWidth={false}
    hintText="e.g. comp 248"
    id="courseSearch"
    listStyle={
      Object {
        "maxHeight": "250px",
        "overflow": "auto",
      }
    }
    menuCloseDelay={300}
    onNewRequest={[Function]}
    onUpdateInput={[Function]}
    open={false}
    openOnFocus={false}
    style={
      Object {
        "marginLeft": "12px",
      }
    }
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  />
</div>
`;
